### PR TITLE
Conversion between GTIN formats

### PR DESCRIPTION
**Pull request by beet**
Dependency for marketplacer/marketplacer#20482

Depends on #4 for updated rubocop settings

Can tell GTIN objects to convert themselves into different formats, while applying validation rules around the code length and check-bit that handle zero-padding.

There are 4 [GTIN standards](https://www.gtin.info) corresponding to the code length:

* GTIN-8
* GTIN-12
* GTIN-13
* GTIN-14

It is valid to convert a subset up to a superset by prefixing it with a zero.

For example, a GTIN-8 code of 12345670 can be represented as a GTIN-12 code of 00001234567.

A superset can be converted back to a subset, only when it is a zero-padded representation. For example, a GTIN-12 code of 00001234567 can be converted back to a GTIN-8 code of 12345670, but a GTIN-12 code of 123445678907 cannot.

## Factory interface

The current public interface to `BarcodeValidation.scan` is maintained, while using a template class structure to return an instance of the GTIN class matching the given barcode's length:

```ruby
gtin = BarcodeValidation.scan("12345670")
=> #<BarcodeValidation::GTIN::GTIN8(12345670)>

gtin.valid?
=> true
```

## Conversion between formats

All GTIN objects provide public instance methods to explicitly convert themselves into one of the 4 GTIN formats:

* `to_gtin_8`
* `to_gtin_12`
* `to_gtin_13`
* `to_gtin_14`

If the GTIN can be converted into that format by either zero-padding, or removing leading zeros, it will return a valid instance of the corresponding GTIN class:

```ruby
gtin = BarcodeValidation.scan("00001234567")
=> #<BarcodeValidation::GTIN::GTIN12(00001234567)>

gtin8 = gtin.to_gtin_8
=> #<BarcodeValidation::GTIN::GTIN8(12345670)>

gtin8.valid?
=> true
```

If it cannot be converted into the given format, it will return a `BarcodeValidation::InvalidGTIN` instance:

```ruby
gtin = BarcodeValidation.scan("123456789012")
=> #<BarcodeValidation::GTIN::GTIN13(123456789012)>

gtin8 = gtin.to_gtin_8
=> #<BarcodeValidation::InvalidGTIN input="123456789012" error="invalid value for BarcodeValidation::GTIN::GTIN8(): "123456789012"">

gtin8.valid?
=> false

gtin8.error_message
=> "invalid value for BarcodeValidation::GTIN::GTIN8(): \"123456789012\""
```

### Converting to all valid formats with `to_all_valid`

For DB lookups in the Marketplacer app, GTIN objects provide the `to_all_valid` method, which returns a collection of GTIN objects for valid formats that the code can be converted into. For example, a GTIN-13 code that is a zero-padded GTIN-12 can be converted down to GTIN-12, and up to GTIN-14, but not down to GTIN-8:

```
gtin = BarcodeValidation.scan("0123456789012")
=> #<BarcodeValidation::GTIN::GTIN13(123456789012)>

gtin.to_all_valid
=> [
  #<BarcodeValidation::GTIN::GTIN12(123456789012)>,
  #<BarcodeValidation::GTIN::GTIN13(0123456789012)>,
  #<BarcodeValidation::GTIN::GTIN14(00123456789012)>
]
```

### CODE REVIEW

**Comment by atrac303** [comment]
it's usual practice to rebase and squash all your commits into one!

File: lib/barcodevalidation/gtin.rb [hunk 1]
```diff
@@ -1,46 +1,20 @@
...
-      def valid?
-        actual == expected
-      end
-
-      def inspect
-        return super if valid?
-
-        "#<#{self.class}(#{actual}) invalid: expected #{expected}>"
-      end
+      module_eval("GTIN#{input.to_s.size}", __FILE__, __LINE__).new(input)
```
Comment (robotdana) [inline lib/barcodevalidation/gtin.rb]: i’d rather this was a boring case statement than used eval.

File: lib/barcodevalidation/gtin/base.rb [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+      end
+
+      # Instantiates the given class with the string value of this instance
+      # with any leading zeros stripped out, and then zero-padded up to the
+      # valid length of the given class. If the resulting string is <> the
+      # valid length of the target format, the returned object will be a
+      # BarcodeValidation::InvalidGTIN with valid? = false and a meaningful
+      # error message.
+      def transcode_to(klass)
+        gtin = klass.new(format("%0#{klass.new(nil).valid_length}d", to_s.gsub(/^0+/, "")))
```
Comment (robotdana) [inline lib/barcodevalidation/gtin/base.rb]: klass.new(nil).valid_length makes me think this should be a class method as well

File: lib/barcodevalidation/gtin.rb (hunk 1 above)
Comment (beet) [reply lib/barcodevalidation/gtin.rb]: Really? Not a fan of case switches, or even conditionals if I can at all avoid them, but OK.

File: lib/barcodevalidation/gtin/base.rb (hunk 2 above)
Comment (beet) [reply lib/barcodevalidation/gtin/base.rb]: Yeah, it did feel like that, will do.

File: lib/barcodevalidation/gtin.rb (hunk 1 above)
Comment (beet) [reply lib/barcodevalidation/gtin.rb]: Heh, tried it, but rubocop's `Metrics/MethodLength` fails with "Method has too many lines. [12/10]" :) 

Bump it up here, or go through another PR for the rubocop settings?

Comment (beet) [reply lib/barcodevalidation/gtin.rb]: Actually, even though I'm not a fan, with 2584faa it would actually look OK: _(just long, and repetitive)_

```ruby
def self.new(input)
  case input.to_s.size
  when GTIN8::VALID_LENGTH
    GTIN8
  when GTIN12::VALID_LENGTH
    GTIN12
  when GTIN13::VALID_LENGTH
    GTIN13
  when GTIN14::VALID_LENGTH
    GTIN14
  else
    BarcodeValidation::InvalidGTIN
  end.new(input)
end
```

It's explicit without using magic numbers, but we'd need to adjust rubocop.

Comment (beet) [reply lib/barcodevalidation/gtin.rb]: Actually, pardon the comment spam, but rubocop is happy with a hash pattern:

```ruby
def self.new(input)
  klass = {
    8 => GTIN8,
    12 => GTIN12,
    13 => GTIN13,
    14 => GTIN14,
  }[input.to_s.size] || BarcodeValidation::InvalidGTIN

  klass.new(input)
end
```

That actually looks neater to me than the case switch, what do you think?

Comment (robotdana) [reply lib/barcodevalidation/gtin.rb]: make the hash a constant and it’s pretty much equivalent in c.

Comment (robotdana) [reply lib/barcodevalidation/gtin.rb]: (i’d tend to use `when then` to avoid the long method but this also works)

Comment (robotdana) [reply lib/barcodevalidation/gtin.rb]: and if possible i tend to prefer not using branches but eval is not something id use in favour of branches

Comment (beet) [reply lib/barcodevalidation/gtin.rb]: OK, hoping what I've just pushed up is a happy medium:

```ruby
def new(input)
  (class_for_input(input) || BarcodeValidation::InvalidGTIN).new(input)
end

private

def class_for_input(input)
  [GTIN8, GTIN12, GTIN13, GTIN14].find do |klass|
    input.to_s.size == klass::VALID_LENGTH
  end
end
```

Not as concise as eval, but also not as verbose as a case switch.

File: lib/barcodevalidation/gtin/base.rb [hunk 3]
```diff
@@ -0,0 +1,102 @@
+# frozen_string_literal: true
+
+require "forwardable"
+
+module BarcodeValidation
+  module GTIN
+    class Base < BarcodeValidation::DigitSequence
+      MODULUS = 10
+
+      extend Forwardable
```
Comment (robotdana) [inline lib/barcodevalidation/gtin/base.rb:10]: i'm not sure where we're using forwardable in this class or its descendants?

File: spec/lib/barcodevalidation_spec.rb [hunk 4]
```diff
@@ -47,7 +47,6 @@
     Class.new,
     Object.new,
     0...10,
-    /abcd/,
```
Comment (robotdana) [inline spec/lib/barcodevalidation_spec.rb]: why did you remove this?

**Review by robotdana — APPROVED** [review]

File: lib/barcodevalidation/gtin/base.rb (hunk 3 above)
Comment (beet) [reply lib/barcodevalidation/gtin/base.rb:10]: Nor am I, I just abstracted this class out of the main `BarcodeValidation` as-is with fe13cc9

File: spec/lib/barcodevalidation_spec.rb (hunk 4 above)
Comment (beet) [reply spec/lib/barcodevalidation_spec.rb]: See 49e8f5f

A Ruby regex is not a valid barcode GTIN and it was breaking my factory when using eval, but now that I've changed it to using an array pattern I could revert that.
